File: Examples/Python/mux_connection.py
import time

import connection
import helpers
import ximu3

# Search for connection
usb_devices = ximu3.PortScanner.scan_filter(ximu3.PORT_TYPE_USB)

time.sleep(1)  # wait for OS to release port

if not usb_devices:
    raise Exception("No USB connections available")

print(f"Found {usb_devices[0]}")

# Open connection
usb_connection = ximu3.Connection(usb_devices[0].connection_config).open()

# Mux connection
if helpers.yes_or_no("Search for connections?"):
    mux_devices = ximu3.MuxScanner.scan(usb_connection)

    if not mux_devices:
        raise Exception("No mux connections available")

    print(f"Found {mux_devices[0]}")

    config = mux_devices[0].connection_config

    connection.run(config)
else:
    config = ximu3.MuxConnectionConfig(0x41, usb_connection)  # replace with actual connection config

    connection.run(config)

# Close connection
usb_connection.close()

File: Examples/CSharp/x-IMU3-Examples/Examples/MuxConnection.cs
namespace Ximu3Examples
{
    class MuxConnection : Connection
    {
        public MuxConnection()
        {
            // Search for connection
            var usbDevices = Ximu3.PortScanner.ScanFilter(Ximu3.CApi.XIMU3_PortType.XIMU3_PortTypeUsb);

            System.Threading.Thread.Sleep(1000); // wait for OS to release port

            if (usbDevices.Length == 0)
            {
                Console.WriteLine("No USB connections available");
                return;
            }

            Console.WriteLine("Found " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_device_to_string(usbDevices[0])));

            // Open connection
            var usbConnection = new Ximu3.Connection(Ximu3.ConnectionConfig.From(usbDevices[0])!);

            var result = usbConnection.Open();

            if (result != Ximu3.CApi.XIMU3_Result.XIMU3_ResultOk)
            {
                Console.WriteLine("Unable to open " + usbConnection.GetConfig() + ". " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_result_to_string(result)) + ".");
                return;
            }

            // Mux connection
            if (Helpers.YesOrNo("Connect automatically?"))
            {
                var muxDevices = Ximu3.MuxScanner.Scan(usbConnection);

                if (muxDevices.Length == 0)
                {
                    Console.WriteLine("No mux connections available");
                    usbConnection.Close();
                    return;
                }

                var config = Ximu3.ConnectionConfig.From(muxDevices[0]);

                Run(config!);
            }
            else
            {
                var config = new Ximu3.MuxConnectionConfig(0x41, usbConnection); // replace with actual connection config

                Run(config);
            }

            // Close connection
            usbConnection.Close();
        }
    }
}
